anndata/_config.py

```diff
@@ -0,0 +1,332 @@
+from __future__ import annotations
+
+import os
+import warnings
+from collections.abc import Iterable
+from contextlib import contextmanager
+from functools import wraps
+from inspect import Parameter, signature
+from typing import TYPE_CHECKING, Any, NamedTuple, TypeVar
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Sequence
+
+T = TypeVar("T")
+
+
+class DeprecatedOption(NamedTuple):
+    option: str
+    message: str | None
+    removal_version: str | None
+
+
+# TODO: inherit from Generic[T] as well after python 3.9 is no longer supported
+class RegisteredOption(NamedTuple):
+    option: str
+    default_value: T
+    doc: str
+    validator: Callable[[T], None] | None
+
+
+def update_override_function(register: Callable):
+    """This function updates the keyword arguments, docstring, and annotations of the `SettingsManager.override` function as the `SettingsManager.register` method is called.
+
+    Parameters
+    ----------
+    register
+        This is a reference to the register function of `SettingsManager`
+
+    Returns
+    -------
+    Callable
+        The wrapped `SettingsManager.register` function.
+    """
+
+    @wraps(register)
+    def _impl(
+        self,
+        option: str,
+        default_value: object,
+        doc: str,
+        validator: Callable[[T], None],
+    ):
+        method_output = register(self, option, default_value, doc, validator)
+        # Update annotations for type checking.
+        self.override.__annotations__[option] = type(default_value)
+        # __signature__ needs to be updated for tab autocompletion in IPython.
+        # See https://github.com/ipython/ipython/issues/11624 for inspiration.
+        setattr(
+            self.override.__func__,  # Attributes of a method are immutable, so need to use __func__.
+            "__signature__",
+            signature(self.override).replace(
```

Comment:
`setattr` is when the `name` parameter isn’t fixed

```suggestion
        # Attributes of a method are immutable, so need to use __func__.
        self.override.__func__.__signature__= signature(self.override).replace(
```

Reply:
Huh, thought I had read this doesn't work because of `__signature__` needing to be initialized specially, but seems to work now.  Great!

Reply:
`x.a = v` is exactly the same as `setattr(x, 'a', v)`.

Both go through `__setattr__` if available and so on.